From 06e127cb81da602410977d33f9c3067f6a9ac2bc Mon Sep 17 00:00:00 2001
From: Julian Greilich <j.greilich@gmx.de>
Date: Mon, 11 May 2026 11:15:30 +0200
Subject: a11y: Expose expandable/expanded in QQuickAccessibleAttached

These properties already exist in QAccessible::State but were not
usable for custom components.

Task-number: QTBUG-146552
Change-Id: Ifb84a618c5975797b4b6434671dbadaa7cb735d7
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
Reviewed-by: Michael Weghorn <m.weghorn@posteo.de>
(cherry picked from commit 477ebeb6dee22eda2e9afc25943700c1c5cc94d4)
---
 src/quick/items/qquickaccessibleattached.cpp  | 14 +++++++
 src/quick/items/qquickaccessibleattached_p.h  |  2 +
 .../qquickaccessible/data/expandable.qml      | 39 +++++++++++++++++++
 .../qquickaccessible/tst_qquickaccessible.cpp | 36 +++++++++++++++++
 4 files changed, 91 insertions(+)
 create mode 100644 tests/auto/quick/qquickaccessible/data/expandable.qml

diff --git x/qtdeclarative/src/quick/items/qquickaccessibleattached.cpp y/qtdeclarative/src/quick/items/qquickaccessibleattached.cpp
index 7f618687731cdb37094f466afb8b57dc835cbae2..6d040d5017bb5775539d1703aeb79894b71b5087 100644
--- x/qtdeclarative/src/quick/items/qquickaccessibleattached.cpp
+++ y/qtdeclarative/src/quick/items/qquickaccessibleattached.cpp
@@ -157,6 +157,20 @@ QT_BEGIN_NAMESPACE
     By default this property is \c false.
     \sa checkable
 */
+/*! \qmlproperty bool QtQuick::Accessible::expandable
+    \brief This property holds whether this item is expandable.
+
+    \since 6.12
+    By default this property is \c false.
+    \sa expanded
+*/
+/*! \qmlproperty bool QtQuick::Accessible::expanded
+    \brief This property holds whether this item is currently expanded.
+
+    \since 6.12
+    By default this property is \c false.
+    \sa expandable
+*/
 /*! \qmlproperty bool QtQuick::Accessible::editable
     \brief This property holds whether this item has editable text.
 
diff --git x/qtdeclarative/src/quick/items/qquickaccessibleattached_p.h y/qtdeclarative/src/quick/items/qquickaccessibleattached_p.h
index a991ab2e269bb3e6f649a58660ca3a1157bbddd3..d1d31033556d6a05f79975bd8bbc7aa02286d527 100644
--- x/qtdeclarative/src/quick/items/qquickaccessibleattached_p.h
+++ y/qtdeclarative/src/quick/items/qquickaccessibleattached_p.h
@@ -70,6 +70,8 @@ public:
     STATE_PROPERTY(checkable)
     STATE_PROPERTY(checked)
     STATE_PROPERTY(editable)
+    STATE_PROPERTY(expandable)
+    STATE_PROPERTY(expanded)
     STATE_PROPERTY(focusable)
     STATE_PROPERTY(focused)
     STATE_PROPERTY(multiLine)
diff --git x/qtdeclarative/tests/auto/quick/qquickaccessible/data/expandable.qml y/qtdeclarative/tests/auto/quick/qquickaccessible/data/expandable.qml
new file mode 100644
index 0000000000000000000000000000000000000000..211abe5254713a07206b7ae89c62c2e92884ac42
--- /dev/null
+++ y/qtdeclarative/tests/auto/quick/qquickaccessible/data/expandable.qml
@@ -0,0 +1,39 @@
+import QtQuick 2.0
+
+Item {
+    width: 400
+    height: 400
+    Accessible.name: "root"
+
+    // button, not expandable
+    Rectangle {
+        objectName: "button1"
+        y: 20
+        width: 100; height: 20
+        Accessible.role : Accessible.Button
+    }
+
+    // button, expandable, not expanded
+    Rectangle {
+        objectName: "button2"
+        y: 40
+        width: 100; height: 20
+        Accessible.role : Accessible.Button
+        Accessible.expandable: expandable
+        Accessible.expanded: expanded
+        property bool expandable: true
+        property bool expanded: false
+    }
+
+    // button, expandable, expanded
+    Rectangle {
+        objectName: "button3"
+        y: 60
+        width: 100; height: 20
+        Accessible.role : Accessible.Button
+        Accessible.expandable: expandable
+        Accessible.expanded: expanded
+        property bool expandable: true
+        property bool expanded: true
+    }
+}
diff --git x/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp y/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp
index 60439ea08557ab8b03fc7a925c2cc23a4f93b9eb..e3306b49bfff1f191ddb17af8afb6323f01e89e6 100644
--- x/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp
+++ y/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp
@@ -64,6 +64,7 @@ private slots:
     void basicPropertiesTest();
     void hitTest();
     void checkableTest();
+    void expandableTest();
     void ignoredTest();
     void passwordTest();
     void announceTest();
@@ -789,6 +790,41 @@ void tst_QQuickAccessible::checkableTest()
     QTestAccessibility::clearEvents();
 }
 
+void tst_QQuickAccessible::expandableTest() {
+    auto window = std::make_unique<QQuickView>();
+    window->setSource(testFileUrl("expandable.qml"));
+    window->show();
+
+    QQuickItem *contentItem = window->contentItem();
+    QVERIFY(contentItem);
+    QQuickItem *rootItem = contentItem->childItems().first();
+    QVERIFY(rootItem);
+
+    QAccessibleInterface *iface =
+        QAccessible::queryAccessibleInterface(window.get());
+    QVERIFY(iface);
+    QAccessibleInterface *root = iface->child(0);
+
+    QAccessibleInterface *button1 = root->child(0);
+    QCOMPARE(button1->role(), QAccessible::Button);
+    QVERIFY(!(button1->state().expanded));
+    QVERIFY(!(button1->state().expandable));
+
+    QAccessibleInterface *button2 = root->child(1);
+    QVERIFY(!(button2->state().expanded));
+    QVERIFY(button2->state().expandable);
+    QQuickItem *button2item =
+        qobject_cast<QQuickItem *>(rootItem->childItems().at(1));
+    QVERIFY(button2item);
+    QCOMPARE(button2item->objectName(), QLatin1String("button2"));
+
+    QAccessibleInterface *button3 = root->child(2);
+    QVERIFY(button3->state().expanded);
+    QVERIFY(button3->state().expandable);
+
+    QTestAccessibility::clearEvents();
+}
+
 void tst_QQuickAccessible::ignoredTest()
 {
     auto window = std::make_unique<QQuickView>();
